src/ColorPicker/Halogen/Component.purs

```diff
@@ -0,0 +1,347 @@
+import Data.Either (Either(..), either, isLeft)
+import Data.Either.Nested as Either
+import Data.Foldable (fold, foldMap, for_)
+import Data.Functor.Coproduct.Nested as Coproduct
+import Data.Map (Map, insert, lookup)
+import Data.Maybe (Maybe(..), fromMaybe)
+import Data.Monoid (mempty)
+import Halogen (liftEff)
+import Halogen as H
+import Halogen.Component.ChildPath as CP
+import Halogen.HTML as HH
+import Halogen.HTML.CSS as HCSS
+import Halogen.HTML.Events as HE
+import Halogen.HTML.Properties as HP
+import Halogen.Query.HalogenM (halt)
+import NumberInput.Halogen.Component as Num
+
+
+type ValueProgress a =  { current ∷ a, next ∷ a }
+type State =
+  { color ∷ ValueProgress Color
+  , inputs ∷ Map String (Either String String)
+  , props ∷ Props
+  }
+
+data Message = NextChange Color | NotifyChange Color
+
+
+type ColorComponentGroups = Array ColorComponents
+type ColorComponents = Array ColorComponent
+
+classesFor ∷ Props → ClassGroup → Array HH.ClassName
+classesFor {classes} key = fromMaybe [] $ lookup key classes
+
+data ClassGroup
+  = Root
+  | Dragger
+  | Field
+  | FieldGradient
+  | FieldSelector
+  | Slider
+  | SliderSelector
+  | Aside
+  | Stage
+  | ColorBlockCurrent
+  | ColorBlockNext
+  | Editing
+  | EditingItem
+  | Input
+  | InputLabel
+  | InputElem
+  | InputElemInvalid
+  | Actions
+  | ActionSet
+
+derive instance classGroupEq ∷ Eq ClassGroup
+derive instance classGroupOrd ∷ Ord ClassGroup
+
+type Props =
+  { classes ∷ Map ClassGroup (Array HH.ClassName)
+  , editing ∷ ColorComponentGroups
+  }
+
+
+data Query a
+  = SetProps Props a
+  | FieldDragStart Drag.CursorEvent a
+  | FieldDragMove Drag.DragEvent a
+  | SliderDragStart Drag.CursorEvent a
+  | SliderDragMove Drag.DragEvent a
+  | ComponentUpdate (Color → Maybe Color) a
+  | Commit a
+  | Init a
+  | GetValue (ValueProgress Color → a)
+  | SetValue (ValueProgress Color) a
+  | UpdateTextInput String String (Maybe Color) a
+
+type ChildQuery = Coproduct.Coproduct1 (Num.Query Number)
+type Slot = Either.Either1 String
+
+cpNumComponent ∷ CP.ChildPath (Num.Query Number) ChildQuery String Slot
+cpNumComponent = CP.cp1
+
+
+type HTML m = H.ParentHTML Query ChildQuery Slot m
+
+type DSL m = H.ParentDSL State Query ChildQuery Slot Message m
+
+type PickerEffects r = Drag.DragEffects r
+
+initialColor ∷ Color
+initialColor = Color.hsl 0.0 0.0 0.0
+
+picker ∷ ∀ m r. MonadAff (PickerEffects r) m ⇒ H.Component HH.HTML Query Props Message m
+picker = H.lifecycleParentComponent
+  { initialState: { color: { current: initialColor, next: initialColor }, inputs: mempty, props: _ }
+  , render
+  , eval
+  , receiver: HE.input SetProps
+  , initializer: Just $ H.action Init
+  , finalizer: Nothing
+  }
+
+render ∷ ∀ m. State → HTML m
+render state@{ color, inputs, props} =
+  HH.div
+    [ HP.classes $ props `classesFor` Root ]
+    [ dragger, aside ]
+  where
+  textColor c = CSS.color if Color.isLight c then CSS.black else CSS.white
```

Comment:
might want to do this through a class. White or Black are quite harsh and it seems very jumpy when the UI switches between the two color states.